### PR TITLE
add multi data source support

### PR DESCRIPTION
**Pull request by wanglam**
### Description
1. Add multi data source support for admin UI
2. Add missing unit tests for server side
image

image

### Issues Resolved
_List any issues this PR will resolve, e.g. Closes [...]._ 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: public/components/app.tsx [hunk 1]
```diff
@@ -38,27 +51,49 @@ export const MlCommonsPluginApp = ({
   http,
   chrome,
   data,
+  dataSource,
+  dataSourceManagement,
+  savedObjects,
+  setActionMenu,
 }: MlCommonsPluginAppDeps) => {
+  const dataSourceEnabled = !!dataSource?.dataSourceEnabled;
```
Comment (BionIT) [inline public/components/app.tsx:59]: nit: we should be able to tell if the data source plugin is enabled via `!!dataSource`

File: public/components/data_source_top_nav_menu.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+export const DataSourceTopNavMenu = ({
+  dataSource,
+  savedObjects,
+  notifications,
+  setActionMenu,
+  dataSourceManagement,
+}: DataSourceTopNavMenuProps) => {
+  const dataSourceEnabled = !!dataSource?.dataSourceEnabled;
```
Comment (BionIT) [inline public/components/data_source_top_nav_menu.tsx:31]: This component is only mounted when the dataSourceEnabled flag is true, do we need to double check the flag?

Comment (wanglam) [reply public/components/data_source_top_nav_menu.tsx:31]: Do we mean the `dataSourceEnabled` will always be true, if the top nav menu mounted? Add check here is for `dataSourceEnabled` changed. The `dataSourceEnabled` in DataSourceContextProvider will be updated if enabled status changed from outside.

File: public/components/app.tsx (hunk 1 above)
Comment (wanglam) [reply public/components/app.tsx:59]: Do you mean we don't need to read the `dataSourceEnabled` property in `dataSource`?  The multi data source will always be enabled if `dataSource` is truly.

**Review by BionIT — APPROVED** [review]

File: public/hooks/use_fetcher.ts [hunk 3]
```diff
@@ -5,13 +5,20 @@
 
 import { useCallback, useEffect, useRef, useState } from 'react';
 
+export const DO_NOT_FETCH = Symbol('DO_NOT_FETCH');
```
[the remark this answers is not in the document]
Comment (wanglam) [reply public/hooks/use_fetcher.ts:8]: The fetcher will be automatic execute if Component mounted. Add this `DO_NOT_FETCH` here is for prevent this default behavior. If this parameter present, the fetcher won't be executed. In the real world, we don't want any API be execute until data source id finalized.

**Comment by ruanyl** [comment]
I'm still concerning the current implementation as it requires the components to pass data source id explicitly. This is a bit overwhelming if we need to do this in the future for all the components we add that interact with apis. 

I think data source is data layer configuration, presentation layer shouldn't be aware of that. How do you like this idea:

1. since we need UI to reload when data source changed without refreshing the page, we can simply call reload in `useFetcher` when data source changed.
```diff
+ export const currentDataSourceId$ = new BehaviorSubject<string>('');
export const useFetcher = (...) => {
+  const dataSourceId = useObservable(currentDataSourceId$);

+  useEffect(() => {
+    if (dataSourceId) {
+      reload();
 +   }
+  }, [currentDataSourceId$, reload]);
}
```

2. For apis, we could add the data source id by default
```diff
export class Connector {
  public getAll({ dataSourceId }: { dataSourceId?: string }) {
    return InnerHttpProvider.getHttp().get<GetAllConnectorResponse>(CONNECTOR_API_ENDPOINT, {
      query: {
+        data_source_id: currentDataSourceId$.value,
      },
    });
  }
}
```

Then basically, no components need to be changed, and once data source id changed, it reload the data and component get updated.

**Review by ruanyl — APPROVED** [review]

**Comment by wanglam** [comment]
> I'm still concerning the current implementation as it requires the components to pass data source id explicitly. This is a bit overwhelming if we need to do this in the future for all the components we add that interact with apis.
> 
> I think data source is data layer configuration, presentation layer shouldn't be aware of that. How do you like this idea:
> 
> 1. since we need UI to reload when data source changed without refreshing the page, we can simply call reload in `useFetcher` when data source changed.
> 
> ```diff
> + export const currentDataSourceId$ = new BehaviorSubject<string>('');
> export const useFetcher = (...) => {
> +  const dataSourceId = useObservable(currentDataSourceId$);
> 
> +  useEffect(() => {
> +    if (dataSourceId) {
> +      reload();
>  +   }
> +  }, [currentDataSourceId$, reload]);
> }
> ```
> 
> 2. For apis, we could add the data source id by default
> 
> ```diff
> export class Connector {
>   public getAll() {
>     return InnerHttpProvider.getHttp().get<GetAllConnectorResponse>(CONNECTOR_API_ENDPOINT, {
>       query: {
> +        data_source_id: currentDataSourceId$.value,
>       },
>     });
>   }
> }
> ```
> 
> Then basically, no components need to be changed, and once data source id changed, it reload the data and component get updated.

This is a good idea, especially for model registry. There is still one question need to be figure out after move automatic reload logic to `useFetcher`. Sometimes the search params need to updated imediality after data source change. That will cause two requests send. One triggered by automatic reload, another triggered by params change. The `useFetcher` hook should cancel the previous request. The `useFetcher` doesn't support this now. I've raised a refactor [issue](https://github.com/opensearch-project/ml-commons-dashboards/issues/316) to track it. Will refactor this part in model registry.

**Review by raintygao — APPROVED** [review]

**Pull request merged**